# Week 7: Instructor Led Lab
Author: Jen Estes \
Course: BGEN 632 - Introduction to Python \
Term: Spring 2025 \
Due Date: April 14th, 2025 

This notebook contains code for the inspection and organization of data based on the requirements outlined by Dr. Newton in the BGEN632 Week 7 GitHub repo. While using  the github_teams.csv file, the program primarily makes use of pandas DataFrames to do this analysis.

----

### Importing Modules

In [348]:
import os
import pandas as pd
import numpy as np
from sklearn.model_selection import KFold

### Setting Working Directory

In [351]:
os.getcwd()  # get current working directory, note that data is in "data" folder

'/Users/jenestes/Desktop/week7labs/data'

In [353]:
# want working directory to be within this data folder
os.chdir("/Users/jenestes/Desktop/week7labs/data")   # change the directory
os.getcwd()                                          # confirm the change 

'/Users/jenestes/Desktop/week7labs/data'

## Accessing Data

### Load Data

In [357]:
gh_teams = pd.read_csv("github_teams.csv")

### Inspect Columns
This method gives us more information than just the column headers, but this is useful for inspecting the data frame.

In [360]:
%%time

gh_teams.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 608 entries, 0 to 607
Data columns (total 19 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   name_h                    608 non-null    object 
 1   Team_type                 608 non-null    object 
 2   Team_size_class           608 non-null    object 
 3   human_members_count       608 non-null    int64  
 4   bot_members_count         608 non-null    int64  
 5   human_work                608 non-null    int64  
 6   work_per_human            608 non-null    float64
 7   human_gini                608 non-null    float64
 8   human_Push                608 non-null    int64  
 9   human_IssueComments       608 non-null    int64  
 10  human_PRReviewComment     608 non-null    int64  
 11  human_MergedPR            608 non-null    int64  
 12  bot_work                  608 non-null    int64  
 13  bot_Push                  608 non-null    int64  
 14  bot_IssueC

### Number of Columns and Rows
The output above suggests there are 608 rows and 19 columns, but we can confirm this using the shape function, which will output the (# of rows, # of columns).

In [363]:
gh_teams.shape

(608, 19)

### Converting Columns
Our initial output using the info function shows that the columns that are categorical are name_h, Team_type, and Team_size_class. We can convert these from *object* to *category*. We can confirm this by using the info function again, after these changes are made.

In [366]:
gh_teams['name_h'] = gh_teams['name_h'].astype('category')               # making changes
gh_teams['Team_type'] = gh_teams['Team_type'].astype('category')            # making changes
gh_teams['Team_size_class'] = gh_teams['Team_size_class'].astype('category')  # making changes
gh_teams.info()                                                                 # confirm the changes

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 608 entries, 0 to 607
Data columns (total 19 columns):
 #   Column                    Non-Null Count  Dtype   
---  ------                    --------------  -----   
 0   name_h                    608 non-null    category
 1   Team_type                 608 non-null    category
 2   Team_size_class           608 non-null    category
 3   human_members_count       608 non-null    int64   
 4   bot_members_count         608 non-null    int64   
 5   human_work                608 non-null    int64   
 6   work_per_human            608 non-null    float64 
 7   human_gini                608 non-null    float64 
 8   human_Push                608 non-null    int64   
 9   human_IssueComments       608 non-null    int64   
 10  human_PRReviewComment     608 non-null    int64   
 11  human_MergedPR            608 non-null    int64   
 12  bot_work                  608 non-null    int64   
 13  bot_Push                  608 non-null    int64   

### Unique Values

In [370]:
# Unique values for Team Type - 2
pd.unique(gh_teams.Team_type)  

['human-bot', 'human']
Categories (2, object): ['human', 'human-bot']

In [372]:
# Unique values for Team Size Class - 3
pd.unique(gh_teams.Team_size_class)  

['Small', 'Large', 'Medium']
Categories (3, object): ['Large', 'Medium', 'Small']

### Indexing

In [375]:
# Value of the 63rd row and 6th column
gh_teams.iloc[62,5]

35

In [377]:
# Values for the 300th row
gh_teams.iloc[299]        

name_h                      IyfocAGfAHLncCVJUujqTA/A_QZ6HlUb5sRQHhPa7SGzQ
Team_type                                                       human-bot
Team_size_class                                                    Medium
human_members_count                                                     4
bot_members_count                                                       1
human_work                                                           1049
work_per_human                                                     262.25
human_gini                                                       0.448761
human_Push                                                            739
human_IssueComments                                                   213
human_PRReviewComment                                                  91
human_MergedPR                                                          6
bot_work                                                               52
bot_Push                              

### Three Methods: row with index value 595 with 1st, 2nd, 3rd columns

In [380]:
# Method 1
gh_teams.iloc[595, 0:3]  

name_h             zAh1NECRCquqUJ_-1d6hAw/DET3jTK8hokYfY_neJ1IVQ
Team_type                                              human-bot
Team_size_class                                            Small
Name: 595, dtype: object

In [382]:
# Method 2
gh_teams.iloc[595, :3]  

name_h             zAh1NECRCquqUJ_-1d6hAw/DET3jTK8hokYfY_neJ1IVQ
Team_type                                              human-bot
Team_size_class                                            Small
Name: 595, dtype: object

In [384]:
# Method 3
gh_teams.loc[595, 'name_h':'Team_size_class']

name_h             zAh1NECRCquqUJ_-1d6hAw/DET3jTK8hokYfY_neJ1IVQ
Team_type                                              human-bot
Team_size_class                                            Small
Name: 595, dtype: object

### Two methods: row with index value 46 with the 3rd and 7th columns

In [387]:
# Method 1
gh_teams.iloc[[46], [2, 6]]

Team_size_class  work_per_human
46          Medium       31.833333

In [389]:
# Method 2
gh_teams.loc[46, ["Team_size_class", "work_per_human"]]

Team_size_class       Medium
work_per_human     31.833333
Name: 46, dtype: object

### Two Methods: New DataFrame for the column `bot_work` 

In [392]:
# Method 1 - The column of interest
gh_teams.bot_work

0        43
1         0
2         0
3      1972
4       302
       ... 
603      26
604       0
605       0
606       0
607       8
Name: bot_work, Length: 608, dtype: int64

In [394]:
# Method 1- Creating new data frame
just_bot_work = pd.DataFrame(gh_teams.bot_work)
just_bot_work                                    # confirming that this worked

bot_work
0          43
1           0
2           0
3        1972
4         302
..        ...
603        26
604         0
605         0
606         0
607         8

[608 rows x 1 columns]

In [396]:
# Method 2 - A much longer way to do this 
gh_teams.columns  

Index(['name_h', 'Team_type', 'Team_size_class', 'human_members_count',
       'bot_members_count', 'human_work', 'work_per_human', 'human_gini',
       'human_Push', 'human_IssueComments', 'human_PRReviewComment',
       'human_MergedPR', 'bot_work', 'bot_Push', 'bot_IssueComments',
       'bot_PRReviewComment', 'bot_MergedPR', 'eval_survival_day_median',
       'issues_count'],
      dtype='object')

In [398]:
just_bot_work2 = gh_teams.drop(['name_h', 'Team_type', 'Team_size_class', 'human_members_count',
       'bot_members_count', 'human_work', 'work_per_human', 'human_gini',
       'human_Push', 'human_IssueComments', 'human_PRReviewComment',
       'human_MergedPR', 'bot_Push', 'bot_IssueComments',
       'bot_PRReviewComment', 'bot_MergedPR', 'eval_survival_day_median',
       'issues_count'], axis = 1)
just_bot_work2                                      # confirming that this worked

bot_work
0          43
1           0
2           0
3        1972
4         302
..        ...
603        26
604         0
605         0
606         0
607         8

[608 rows x 1 columns]

## Sorting and Ordering Data

### Looking at Subsets of Data

In [402]:
# `human-bot` teams that have a `bot_members_count` value greater than and equal to 2
gh_teams[(gh_teams.Team_type == 'human-bot') & (gh_teams.bot_members_count >= 2)]

name_h  Team_type Team_size_class  \
3    _l5u7I5p4thtW5SjR_9_4w/aZNCdVXta7fh7eCMzZP1CA  human-bot           Large   
4    _l5u7I5p4thtW5SjR_9_4w/m_FpD7PKQHqVXHn2bh7u2g  human-bot           Large   
42   2-scMrZv13F95YPZmfieww/4Zc56iUYjIZrZU06omFrJw  human-bot           Large   
84   4YoH8row044yJjPIqWJw9Q/NSXj3i61X71lV0StTN71Ww  human-bot           Small   
89   5Is-_ie16OEGmW1arZm8qg/8UeSk2P76pTG7pPLtxsHTQ  human-bot           Large   
110  7sA-8-nyqr0Ri2CT4-FSZw/GJPQoUhHfvUsxKcdkHWLEw  human-bot           Small   
146  bi5TY2Z4OSQq3PMs6JnKYA/5wtZcUUo1XmLHIra8NDtFQ  human-bot          Medium   
147  bi5TY2Z4OSQq3PMs6JnKYA/9b9IqkDK14ketwn88f3hKA  human-bot           Small   
149  bi5TY2Z4OSQq3PMs6JnKYA/kIiAIJpk6lOa6Nxf234KkQ  human-bot           Small   
224  FAhkB4rsocfDW0vrM8U8NA/3KHgTzOwWtAxTXlp_mbqoA  human-bot           Large   
229  FAhkB4rsocfDW0vrM8U8NA/Tl_ZLGwQZrAi-GHyEKl_jA  human-bot           Large   
231  FAhkB4rsocfDW0vrM8U8NA/zHikYkcLoXRCKt2cUrRvvg  human-bot           Large   
268  hXoZRbHPbVxh--funPXSiw/iNU0l_SpKVjGfHOp8vUt8w  human-bot           Small   
325  kpY6vTbIlrk6Nrjilv8zhw/o_e0F16zwdovR3aFgIQipQ  human-bot           Large   
331  LDm8GbdhrDbcBGJF0dR_5g/-L63SLGp3n6MA-y5Di59og  human-bot           Large   
345  MC6oqT7o22Y_rULWJZllfA/MXyVzmYYom7cgybNB0CjFQ  human-bot           Large   
376  O8b2IFn80coqzdP6IwbEKg/uAklmRuq-bJgMfb-PuSjSg  human-bot          Medium   
413  Q658evmRIuaDEnUGlBqzYw/m-hxXTPKmxu3fh32f-RHtg  human-bot           Small   
441  ReeJIVTm1BiV_SO5tNDVFA/ZfIj2C4Weg8to684MM7PGA  human-bot           Large   
472  sq4hB0bA5jL3X3tvK_-stg/5YplqVgfP__pUXTJ9SMoQA  human-bot          Medium   
473  sq4hB0bA5jL3X3tvK_-stg/G5WoDm7Rms4zNbJ3nc6hZA  human-bot           Large   
541  vPPJXk7AxaaZyPsmA3MniQ/euqt0qlPNtbUnKy41316mA  human-bot          Medium   
568  X1pcjhqqelcsRxAhcbNu-w/HJ1fkoQOuwxXJ25L7FGEcw  human-bot          Medium   
586  YQI5l7pwpiRU8KVKXkNjjg/fCxq0R6_z233Fdr2Ly6iug  human-bot          Medium   

     human_members_count  bot_members_count  human_work  work_per_human  \
3                    234                 12       14579       62.303419   
4                     38                  8        1625       42.763158   
42                    23                  2        4648      202.086957   
84                     2                  2         114       57.000000   
89                    17                  4        7412      436.000000   
110                    3                  2         244       81.333333   
146                    4                  2         170       42.500000   
147                    3                  2         189       63.000000   
149                    3                  2          88       29.333333   
224                   15                  2        4821      321.400000   
229                   18                  2        2328      129.333333   
231                   17                  2         816       48.000000   
268                    2                  2         910      455.000000   
325                   29                  2        2891       99.689655   
331                   18                  2         840       46.666667   
345                    7                  2        1421      203.000000   
376                    4                  2          35        8.750000   
413                    2                  2         256      128.000000   
441                   15                  2        2211      147.400000   
472                    4                  2         484      121.000000   
473                   27                  2        2282       84.518519   
541                    4                  2         297       74.250000   
568                    5                  2         281       56.200000   
586                    5                  2        1644      328.800000   

     human_gini  human_Push  human_IssueComments  human_PRReviewComment  \
3      0.738342        1942                11430                   

In [404]:
# `human` teams that are `Large` and have a `human_gini` value greater than and equal to 0.75
gh_teams[(gh_teams.Team_type == 'human') & (gh_teams.Team_size_class == 'Large') & (gh_teams.human_gini >= 0.75) ]

name_h Team_type Team_size_class  \
138  ASYGR96YA91p3z7MNKjZCA/IB2pZ8ygcvNnlxUdysjSFA     human           Large   
285  IiUao8vA_zm_uEIVVLI-Sw/91ya8vlSP8qgwCllH_6BSw     human           Large   
505  uLHPO58cQefwrJUbyhYOKQ/7YWOP8uDEeKDHQMWKqOoYA     human           Large   
582  y8Jw59EHVSrsluSuhR5okg/V5vb074jNkzg4YCKforX1Q     human           Large   

     human_members_count  bot_members_count  human_work  work_per_human  \
138                   12                  0        1655      137.916667   
285                   25                  0        3599      143.960000   
505                   48                  0        5748      119.750000   
582                    8                  0         277       34.625000   

     human_gini  human_Push  human_IssueComments  human_PRReviewComment  \
138    0.799446         793                  684                    178   
285    0.863507        1249                 2350                      0   
505    0.782040        1715                 3891                    142   
582    0.781137         275                    2                      0   

     human_MergedPR  bot_work  bot_Push  bot_IssueComments  \
138               0         0         0                  0   
285               0         0         0                  0   
505               0         0         0                  0   
582               0         0         0                  0   

     bot_PRReviewComment  bot_MergedPR  eval_survival_day_median  issues_count  
138                    0             0                       4.0         190.0  
285                    0             0                       0.0        1245.0  
505                    0             0                       0.0        1200.0  
582                    0             0                       NaN           NaN

### Counting Rows with Certain Conditions

In [407]:
# Count of teams  in the `Small` or `Large` category- 428
# One Method- output just these teams
gh_teams[(gh_teams.Team_size_class == 'Small')| (gh_teams.Team_size_class == 'Large')]


name_h  Team_type Team_size_class  \
0    _1bqaxzCk0sfQaunsjeViQ/RCEZ3CASdLXbstu9y2JQ7Q  human-bot           Small   
1    _9o07rGiC7DFyi-zm91Q0g/VOgMsrjYEwFAq0BY8kHqGQ      human           Small   
2    _DzK53uaZXnAX3WcC0W28g/Epc4QWw5PNBQIIdvopEHDA      human           Large   
3    _l5u7I5p4thtW5SjR_9_4w/aZNCdVXta7fh7eCMzZP1CA  human-bot           Large   
4    _l5u7I5p4thtW5SjR_9_4w/m_FpD7PKQHqVXHn2bh7u2g  human-bot           Large   
..                                             ...        ...             ...   
601  zPjKmela4b_-cQYzslpJLQ/Y8b87jxGXuKYhNDrcC5AuA      human           Large   
602  zro7Xud3Xy2f5CjF55l_jA/GChw8QQ_KUPepXGZGDWicQ  human-bot           Small   
603  zTj5tlMWgotzJmQl7BP8wQ/iQ914_smScbUO8BI9JlE6A  human-bot           Small   
604  zUBexdmYylGGpxiebXm6gg/sJXD2kulWzU35ijdY3SnBQ      human           Small   
607  ZX09uvu0z_JVPVI-XrPniQ/pxiuGX_R6QI2Euvx2GFIFw  human-bot           Small   

     human_members_count  bot_members_count  human_work  work_per_human  \
0                      2                  1          66       33.000000   
1                      2                  0          62       31.000000   
2                      7                  0         211       30.142857   
3                    234                 12       14579       62.303419   
4                     38                  8        1625       42.763158   
..                   ...                ...         ...             ...   
601                   11                  0         851       77.363636   
602                    2                  1          39       19.500000   
603                    3                  1         855      285.000000   
604                    2                  0          63       31.500000   
607                    2                  1          57       28.500000   

     human_gini  human_Push  human_IssueComments  human_PRReviewComment  \
0      0.287879          29                   33                      4   
1      0.467742          62                    0                      0   
2      0.499661         194                   16                      1   
3      0.738342        1942                11430                   1170   
4      0.666607         203                 1270                    152   
..          ...         ...                  ...                    ...   
601    0.267279         740                  111                      0   
602    0.397436          39                    0                      0   
603    0.474854         423                   59                    373   
604    0.436508          63                    0                      0   
607    0.043860          32                   24                      1   

     human_MergedPR  bot_work  bot_Push  bot_IssueComments  \
0                 0        43         0                 43   
1                 0         0         0                  0   
2                 0         0         0                  0   
3                37      1972         0               1972   
4                 0       302         0                302   
..              ...       ...       ...                ...   
601               0         0         0                  0   
602               0        15         0                 15   
603               0        26         0                 26   
604               0         0         0                  0   
607               0         8         3                  4   

     bot_PRReviewComment  bot_MergedPR  eval_survival_day_median  issues_count  
0                      0             0                      87.0           8.0  
1                      0             0                       NaN           NaN  
2                      0             0                      37.0          46.0  
3                      0             0                       1.0        4757.0  
4                      0             0                       2.0         777.0  
..                   ...           ...

In [409]:
#Another method - create subsample of data and count number of rows 
slteams = gh_teams[(gh_teams.Team_size_class == 'Small')| (gh_teams.Team_size_class == 'Large')]
len(slteams)

428

In [411]:
# Teams in the `Small` or `Large` cateogry with a `human_gini` value less than and equal to 0.20- 66
# Using the second method from above, so that just the count of these teams appears
slteams2 = gh_teams[((gh_teams.Team_size_class == 'Small')| (gh_teams.Team_size_class == 'Large')) & (gh_teams.human_gini <= 0.20)]
len(slteams2)

66

In [413]:
# Count of `human-bot` teams in the `Medium` category- 84
humanbotmedium = gh_teams[(gh_teams.Team_size_class == 'Medium')& (gh_teams.Team_type == 'human-bot')]
len(humanbotmedium)

84

### Sampling Data

In [416]:
# Subsample of 50% of the data.
gh_teams_subsample = gh_teams.sample(frac = 0.5, replace = False) 
gh_teams_subsample

name_h  Team_type Team_size_class  \
604  zUBexdmYylGGpxiebXm6gg/sJXD2kulWzU35ijdY3SnBQ      human           Small   
479  t9G9CZ6ZZN3RCh8Gw2LUQA/t9G9CZ6ZZN3RCh8Gw2LUQA  human-bot           Large   
519  V58hL76xAvyUnitZGjXsZg/CnMANWJYuper3o2pVxj5Ew  human-bot           Large   
453  S8Mlts6voOAuRvrLEzUUwQ/6lwvrEheHXLoLzw2ohH5FA      human           Small   
345  MC6oqT7o22Y_rULWJZllfA/MXyVzmYYom7cgybNB0CjFQ  human-bot           Large   
..                                             ...        ...             ...   
103  7aZIQZSTKAYawn7PY_hphA/pfe5MUSnWVWLDA5CI53JSw  human-bot          Medium   
409  PWktOdCWNXXUO5xh29twIg/-RS4PkkkRPxs1qj5Ky0b4A      human           Small   
235  FHn2oD12fabjlrDK8eaU4A/HCHZsZKYZF-iU02YKO4HeA  human-bot          Medium   
507  uLvKXX_6CJ4zG3V_9-fGRw/YGzgdzxRGvwVZcszx5HyCA  human-bot           Small   
206  eau1pOQYwnzWYNwcKXl-qQ/-ZINNXPWvlSznOXnAJOwkg      human           Small   

     human_members_count  bot_members_count  human_work  work_per_human  \
604                    2                  0          63       31.500000   
479                    7                  1         597       85.285714   
519                    7                  1         332       47.428571   
453                    2                  0          26       13.000000   
345                    7                  2        1421      203.000000   
..                   ...                ...         ...             ...   
103                    6                  1         470       78.333333   
409                    2                  0          36       18.000000   
235                    6                  1         461       76.833333   
507                    3                  1         193       64.333333   
206                    2                  0          27       13.500000   

     human_gini  human_Push  human_IssueComments  human_PRReviewComment  \
604    0.436508          63                    0                      0   
479    0.498684          90                  299                    208   
519    0.512048         213                   91                     28   
453    0.384615          26                    0                      0   
345    0.504072         444                  277                    644   
..          ...         ...                  ...                    ...   
103    0.536170          72                  245                    153   
409    0.305556          36                    0                      0   
235    0.442878         175                  227                     50   
507    0.542314          70                  101                     22   
206    0.425926          24                    3                      0   

     human_MergedPR  bot_work  bot_Push  bot_IssueComments  \
604               0         0         0                  0   
479               0        11         0                 11   
519               0        24         0                 24   
453               0         0         0                  0   
345              56        77         0                 77   
..              ...       ...       ...                ...   
103               0        55         0                 55   
409               0         0         0                  0   
235               9         6         0                  0   
507               0        40         0                 40   
206               0         0         0                  0   

     bot_PRReviewComment  bot_MergedPR  eval_survival_day_median  issues_count  
604                    0             0                       NaN           NaN  
479                    0             0                      57.0         109.0  
519                    0             0                      20.0          26.0  
453                    0             0                       NaN           NaN  
345                    0             0                      19.0         124.0  
..                   ...           ...

In [418]:
#Samples for a 8-fold cross validation test
kf = KFold(n_splits = 8)  # K-Fold cross-validator with 8 folds (the default is 5)

for train, test in kf.split(gh_teams):
    print("%s %s" % (train, test))

[ 76  77  78  79  80  81  82  83  84  85  86  87  88  89  90  91  92  93
  94  95  96  97  98  99 100 101 102 103 104 105 106 107 108 109 110 111
 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126 127 128 129
 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144 145 146 147
 148 149 150 151 152 153 154 155 156 157 158 159 160 161 162 163 164 165
 166 167 168 169 170 171 172 173 174 175 176 177 178 179 180 181 182 183
 184 185 186 187 188 189 190 191 192 193 194 195 196 197 198 199 200 201
 202 203 204 205 206 207 208 209 210 211 212 213 214 215 216 217 218 219
 220 221 222 223 224 225 226 227 228 229 230 231 232 233 234 235 236 237
 238 239 240 241 242 243 244 245 246 247 248 249 250 251 252 253 254 255
 256 257 258 259 260 261 262 263 264 265 266 267 268 269 270 271 272 273
 274 275 276 277 278 279 280 281 282 283 284 285 286 287 288 289 290 291
 292 293 294 295 296 297 298 299 300 301 302 303 304 305 306 307 308 309
 310 311 312 313 314 315 316 317 318 319 320 321 32

In [420]:
gh_teams.iloc[train]

name_h  Team_type Team_size_class  \
0    _1bqaxzCk0sfQaunsjeViQ/RCEZ3CASdLXbstu9y2JQ7Q  human-bot           Small   
1    _9o07rGiC7DFyi-zm91Q0g/VOgMsrjYEwFAq0BY8kHqGQ      human           Small   
2    _DzK53uaZXnAX3WcC0W28g/Epc4QWw5PNBQIIdvopEHDA      human           Large   
3    _l5u7I5p4thtW5SjR_9_4w/aZNCdVXta7fh7eCMzZP1CA  human-bot           Large   
4    _l5u7I5p4thtW5SjR_9_4w/m_FpD7PKQHqVXHn2bh7u2g  human-bot           Large   
..                                             ...        ...             ...   
527  vD1mLoQ_CzlsnXcM1E_WOA/XM83BYKoM4QtXWStPYssrw  human-bot           Small   
528  vD1mLoQ_CzlsnXcM1E_WOA/Yk03OqUO6JLk05uYj9RK6Q  human-bot           Small   
529  VemkfETAeqVyhw0s77AlUw/ERsm7IP5zmTUg2sWF5b-cQ  human-bot           Large   
530  VGYUfaNcvjujHdwS_xv9xA/dts2QrkRvgHdrxZCJYXG6w  human-bot           Small   
531  vihAQ-l2aJ5v4UYrXYRGiw/DEmu3nlEY-yVVBpaZ4dG_Q      human           Large   

     human_members_count  bot_members_count  human_work  work_per_human  \
0                      2                  1          66       33.000000   
1                      2                  0          62       31.000000   
2                      7                  0         211       30.142857   
3                    234                 12       14579       62.303419   
4                     38                  8        1625       42.763158   
..                   ...                ...         ...             ...   
527                    2                  1         106       53.000000   
528                    2                  1          50       25.000000   
529                   14                  1        1878      134.142857   
530                    2                  1          90       45.000000   
531                    7                  0          91       13.000000   

     human_gini  human_Push  human_IssueComments  human_PRReviewComment  \
0      0.287879          29                   33                      4   
1      0.467742          62                    0                      0   
2      0.499661         194                   16                      1   
3      0.738342        1942                11430                   1170   
4      0.666607         203                 1270                    152   
..          ...         ...                  ...                    ...   
527    0.433962          99                    7                      0   
528    0.440000          48                    2                      0   
529    0.753461        1409                  465                      4   
530    0.244444          36                   32                     22   
531    0.536892          62                   28                      1   

     human_MergedPR  bot_work  bot_Push  bot_IssueComments  \
0                 0        43         0                 43   
1                 0         0         0                  0   
2                 0         0         0                  0   
3                37      1972         0               1972   
4                 0       302         0                302   
..              ...       ...       ...                ...   
527               0        18         0                 18   
528               0        10         0                 10   
529               0       597         0                597   
530               0        12         0                 12   
531               0         0         0                  0   

     bot_PRReviewComment  bot_MergedPR  eval_survival_day_median  issues_count  
0                      0             0                      87.0           8.0  
1                      0             0                       NaN           NaN  
2                      0             0                      37.0          46.0  
3                      0             0                       1.0        4757.0  
4                      0             0                       2.0         777.0  
..                   ...           ...

In [422]:
gh_teams.iloc[test]

name_h  Team_type Team_size_class  \
532  VjnX-IGsAsRz8EJKMp6LLA/HQ9CHS_vbj1CSd6OM0aCoA  human-bot           Large   
533  VjnX-IGsAsRz8EJKMp6LLA/qZSqyEmsJJ_Rr7f9toRbwQ  human-bot          Medium   
534  vjYCi8YxMpUj_LaqRdiCXw/wPCNTC9mtdWb8MKJVr439g      human           Large   
535  vlLrA8LGOcUxkQuGbs4TqA/LbQfqlh-Ihko3_Yii02dhQ      human          Medium   
536  vpAJthlySeoTSTCzS0iH9w/co9Uzr_rNRVxqxS0x1UpqA  human-bot          Medium   
..                                             ...        ...             ...   
603  zTj5tlMWgotzJmQl7BP8wQ/iQ914_smScbUO8BI9JlE6A  human-bot           Small   
604  zUBexdmYylGGpxiebXm6gg/sJXD2kulWzU35ijdY3SnBQ      human           Small   
605  zVSBi-iRKCzLiqFwVt6hbg/8SfUBIOeWUjDoQxeUCX7wQ      human          Medium   
606  zVSBi-iRKCzLiqFwVt6hbg/fm_lDWwc8Uu-aZ24BjUNZg      human          Medium   
607  ZX09uvu0z_JVPVI-XrPniQ/pxiuGX_R6QI2Euvx2GFIFw  human-bot           Small   

     human_members_count  bot_members_count  human_work  work_per_human  \
532                   12                  1        3609      300.750000   
533                    6                  1         559       93.166667   
534                   28                  0       10705      382.321429   
535                    5                  0         657      131.400000   
536                    4                  1         143       35.750000   
..                   ...                ...         ...             ...   
603                    3                  1         855      285.000000   
604                    2                  0          63       31.500000   
605                    5                  0          26        5.200000   
606                    5                  0          13        2.600000   
607                    2                  1          57       28.500000   

     human_gini  human_Push  human_IssueComments  human_PRReviewComment  \
532    0.703773        2964                  559                     76   
533    0.710495         534                   17                      2   
534    0.743961        4230                 2337                   3997   
535    0.264231         483                  134                     40   
536    0.456294         119                   10                     14   
..          ...         ...                  ...                    ...   
603    0.474854         423                   59                    373   
604    0.436508          63                    0                      0   
605    0.446154          19                    5                      2   
606    0.246154           8                    4                      1   
607    0.043860          32                   24                      1   

     human_MergedPR  bot_work  bot_Push  bot_IssueComments  \
532              10        34         0                 24   
533               6        46         0                 46   
534             141         0         0                  0   
535               0         0         0                  0   
536               0        95        94                  1   
..              ...       ...       ...                ...   
603               0        26         0                 26   
604               0         0         0                  0   
605               0         0         0                  0   
606               0         0         0                  0   
607               0         8         3                  4   

     bot_PRReviewComment  bot_MergedPR  eval_survival_day_median  issues_count  
532                    0            10                      30.0         518.0  
533                    0             0                      36.0          40.0  
534                    0             0                       4.0          21.0  
535                    0             0                       3.0          74.0  
536                    0             0                       NaN           NaN  
..                   ...           ...

In [424]:
# New DataFrame for numeric columns
# Since there are both integers and floats, exclude the category column
gh_teams_numeric = gh_teams.select_dtypes(exclude = ['category'])
gh_teams_numeric

human_members_count  bot_members_count  human_work  work_per_human  \
0                      2                  1          66       33.000000   
1                      2                  0          62       31.000000   
2                      7                  0         211       30.142857   
3                    234                 12       14579       62.303419   
4                     38                  8        1625       42.763158   
..                   ...                ...         ...             ...   
603                    3                  1         855      285.000000   
604                    2                  0          63       31.500000   
605                    5                  0          26        5.200000   
606                    5                  0          13        2.600000   
607                    2                  1          57       28.500000   

     human_gini  human_Push  human_IssueComments  human_PRReviewComment  \
0      0.287879          29                   33                      4   
1      0.467742          62                    0                      0   
2      0.499661         194                   16                      1   
3      0.738342        1942                11430                   1170   
4      0.666607         203                 1270                    152   
..          ...         ...                  ...                    ...   
603    0.474854         423                   59                    373   
604    0.436508          63                    0                      0   
605    0.446154          19                    5                      2   
606    0.246154           8                    4                      1   
607    0.043860          32                   24                      1   

     human_MergedPR  bot_work  bot_Push  bot_IssueComments  \
0                 0        43         0                 43   
1                 0         0         0                  0   
2                 0         0         0                  0   
3                37      1972         0               1972   
4                 0       302         0                302   
..              ...       ...       ...                ...   
603               0        26         0                 26   
604               0         0         0                  0   
605               0         0         0                  0   
606               0         0         0                  0   
607               0         8         3                  4   

     bot_PRReviewComment  bot_MergedPR  eval_survival_day_median  issues_count  
0                      0             0                      87.0           8.0  
1                      0             0                       NaN           NaN  
2                      0             0                      37.0          46.0  
3                      0             0                       1.0        4757.0  
4                      0             0                       2.0         777.0  
..                   ...           ...                       ...           ...  
603                    0             0                       NaN           NaN  
604                    0             0                       NaN           NaN  
605                    0             0                       NaN           NaN  
606                    0             0                       8.0          10.0  
607                    1             0                       5.0          26.0  

[608 rows x 16 columns]

### Removing Columns, Saving as New DataFrame

In [427]:
# Removing the columns `bot_PRReviewComment` and `bot_MergedPR` from the DataFrame.
# Checking columns first
gh_teams_numeric.columns

Index(['human_members_count', 'bot_members_count', 'human_work',
       'work_per_human', 'human_gini', 'human_Push', 'human_IssueComments',
       'human_PRReviewComment', 'human_MergedPR', 'bot_work', 'bot_Push',
       'bot_IssueComments', 'bot_PRReviewComment', 'bot_MergedPR',
       'eval_survival_day_median', 'issues_count'],
      dtype='object')

In [429]:
# Removing columns
gh_teams_numeric.drop(['bot_PRReviewComment', 'bot_MergedPR'], axis = 1)

human_members_count  bot_members_count  human_work  work_per_human  \
0                      2                  1          66       33.000000   
1                      2                  0          62       31.000000   
2                      7                  0         211       30.142857   
3                    234                 12       14579       62.303419   
4                     38                  8        1625       42.763158   
..                   ...                ...         ...             ...   
603                    3                  1         855      285.000000   
604                    2                  0          63       31.500000   
605                    5                  0          26        5.200000   
606                    5                  0          13        2.600000   
607                    2                  1          57       28.500000   

     human_gini  human_Push  human_IssueComments  human_PRReviewComment  \
0      0.287879          29                   33                      4   
1      0.467742          62                    0                      0   
2      0.499661         194                   16                      1   
3      0.738342        1942                11430                   1170   
4      0.666607         203                 1270                    152   
..          ...         ...                  ...                    ...   
603    0.474854         423                   59                    373   
604    0.436508          63                    0                      0   
605    0.446154          19                    5                      2   
606    0.246154           8                    4                      1   
607    0.043860          32                   24                      1   

     human_MergedPR  bot_work  bot_Push  bot_IssueComments  \
0                 0        43         0                 43   
1                 0         0         0                  0   
2                 0         0         0                  0   
3                37      1972         0               1972   
4                 0       302         0                302   
..              ...       ...       ...                ...   
603               0        26         0                 26   
604               0         0         0                  0   
605               0         0         0                  0   
606               0         0         0                  0   
607               0         8         3                  4   

     eval_survival_day_median  issues_count  
0                        87.0           8.0  
1                         NaN           NaN  
2                        37.0          46.0  
3                         1.0        4757.0  
4                         2.0         777.0  
..                        ...           ...  
603                       NaN           NaN  
604                       NaN           NaN  
605                       NaN           NaN  
606                       8.0          10.0  
607                       5.0          26.0  

[608 rows x 14 columns]

In [431]:
# Saving this as new Data Frame in case we want this
gh_teams_numeric_mod = gh_teams_numeric.drop(['bot_PRReviewComment', 'bot_MergedPR'], axis = 1)
gh_teams_numeric_mod

human_members_count  bot_members_count  human_work  work_per_human  \
0                      2                  1          66       33.000000   
1                      2                  0          62       31.000000   
2                      7                  0         211       30.142857   
3                    234                 12       14579       62.303419   
4                     38                  8        1625       42.763158   
..                   ...                ...         ...             ...   
603                    3                  1         855      285.000000   
604                    2                  0          63       31.500000   
605                    5                  0          26        5.200000   
606                    5                  0          13        2.600000   
607                    2                  1          57       28.500000   

     human_gini  human_Push  human_IssueComments  human_PRReviewComment  \
0      0.287879          29                   33                      4   
1      0.467742          62                    0                      0   
2      0.499661         194                   16                      1   
3      0.738342        1942                11430                   1170   
4      0.666607         203                 1270                    152   
..          ...         ...                  ...                    ...   
603    0.474854         423                   59                    373   
604    0.436508          63                    0                      0   
605    0.446154          19                    5                      2   
606    0.246154           8                    4                      1   
607    0.043860          32                   24                      1   

     human_MergedPR  bot_work  bot_Push  bot_IssueComments  \
0                 0        43         0                 43   
1                 0         0         0                  0   
2                 0         0         0                  0   
3                37      1972         0               1972   
4                 0       302         0                302   
..              ...       ...       ...                ...   
603               0        26         0                 26   
604               0         0         0                  0   
605               0         0         0                  0   
606               0         0         0                  0   
607               0         8         3                  4   

     eval_survival_day_median  issues_count  
0                        87.0           8.0  
1                         NaN           NaN  
2                        37.0          46.0  
3                         1.0        4757.0  
4                         2.0         777.0  
..                        ...           ...  
603                       NaN           NaN  
604                       NaN           NaN  
605                       NaN           NaN  
606                       8.0          10.0  
607                       5.0          26.0  

[608 rows x 14 columns]

In [433]:
# Saving the columns `Team_size_class` and `human_members_count` as a new DataFrame.
gh_teams.columns
gh_teams_final = gh_teams.drop(['name_h', 'Team_type',
       'bot_members_count', 'human_work', 'work_per_human', 'human_gini',
       'human_Push', 'human_IssueComments', 'human_PRReviewComment',
       'human_MergedPR', 'bot_work', 'bot_Push', 'bot_IssueComments',
       'bot_PRReviewComment', 'bot_MergedPR', 'eval_survival_day_median',
       'issues_count'], axis = 1)
gh_teams_final

Team_size_class  human_members_count
0             Small                    2
1             Small                    2
2             Large                    7
3             Large                  234
4             Large                   38
..              ...                  ...
603           Small                    3
604           Small                    2
605          Medium                    5
606          Medium                    5
607           Small                    2

[608 rows x 2 columns]

### Renaming Columns

In [436]:
# Used this method since only 2 columns in DataFrame
gh_teams_final.columns = ['Size_of_team', 'Num_humans']
gh_teams_final

Size_of_team  Num_humans
0          Small           2
1          Small           2
2          Large           7
3          Large         234
4          Large          38
..           ...         ...
603        Small           3
604        Small           2
605       Medium           5
606       Medium           5
607        Small           2

[608 rows x 2 columns]